zcash_client_sqlite/src/wallet/init/migrations/transparent_gap_limit_handling.rs

```diff
@@ -0,0 +1,492 @@
+                WalletMigrationError::CorruptedData(format!(
+                    "Invalid UIVK encoding {}: {}",
+                    uivk_str, e
+                ))
+            })
+        };
+
+        let external_scope_code = KeyScope::EXTERNAL.encode();
+
+        conn.execute_batch(&format!(
+            r#"
+            ALTER TABLE addresses ADD COLUMN key_scope INTEGER NOT NULL DEFAULT {external_scope_code};
+            ALTER TABLE addresses ADD COLUMN transparent_child_index INTEGER;
+            "#
+        ))?;
+
+        #[cfg(feature = "transparent-inputs")]
+        {
+            // If the diversifier index is in the valid range of non-hardened child indices, set
+            // `transparent_child_index` so that we can use it for gap limit handling.
+            let mut di_query = conn.prepare(
+                r#"
+                SELECT account_id, accounts.uivk AS uivk, diversifier_index_be
+                FROM addresses
+                JOIN accounts ON accounts.id = account_id
+                GROUP BY account_id, uivk, diversifier_index_be
+                "#,
+            )?;
+            let mut rows = di_query.query([])?;
+            while let Some(row) = rows.next()? {
+                let account_id: i64 = row.get("account_id")?;
+                let uivk = decode_uivk(row.get("uivk")?)?;
+                let di_be: Vec<u8> = row.get("diversifier_index_be")?;
+                let diversifier_index = decode_diversifier_index_be(&di_be)?;
+
+                let transparent_external = NonHardenedChildIndex::try_from(diversifier_index)
+                    .ok()
+                    .and_then(|idx| {
+                        uivk.transparent()
+                            .as_ref()
+                            .and_then(|external_ivk| external_ivk.derive_address(idx).ok())
+                            .map(|t_addr| (idx, t_addr))
+                    });
+
+                // Add transparent address index metadata and the transparent address corresponding
+                // to the index to the addresses table. We unconditionally set the cached
+                // transparent receiver address in order to simplify gap limit handling; even if a
+                // unified address is generated without a transparent receiver, we still assume
+                // that a transparent-only wallet for which we have imported the seed may have
+                // generated an address at that index.
+                if let Some((idx, t_addr)) = transparent_external {
+                    conn.execute(
+                        r#"
+                        UPDATE addresses
+                        SET transparent_child_index = :transparent_child_index,
+                            cached_transparent_receiver_address = :t_addr
+                        WHERE account_id = :account_id
+                        AND diversifier_index_be = :diversifier_index_be
+                        AND key_scope = :external_scope_code
```

Comment:
Not technically required as we just added this column so everything has external scope.